platforms/jvm/plugins-java/src/main/java/org/gradle/api/plugins/JavaPlugin.java

```diff
@@ -256,7 +261,11 @@ public void apply(final Project project) {
         project.getPluginManager().apply("org.gradle.jvm-test-suite"); // TODO: change to reference plugin class by name after project dependency cycles untangled; this will affect ApplyPluginBuildOperationIntegrationTest (will have to remove id)
 
         JavaPluginExtension javaExtension = project.getExtensions().getByType(JavaPluginExtension.class);
-        DefaultJvmSoftwareComponent javaComponent = createJavaComponent(project, javaExtension);
+
+        JvmFeatureInternal mainFeature = createMainFeature(projectInternal, javaExtension);
+        JvmTestSuite defaultTestSuite = configureBuiltInTest(mainFeature, projectInternal.getConfigurations(), project.getTasks(), project.getExtensions(), objectFactory);
+        JvmSoftwareComponentInternal javaComponent = createJavaComponent(mainFeature, defaultTestSuite, project.getObjects(), project.getComponents());
```

Comment:
🤔 I think creating the main feature and the default test suite should be moved inside the `createJavaComponent` method.  Then the mainFeature can be retrieved via the component for configuring diagnostics below.  Applying the plugin/this method isn't the right level of abstraction to consider every last detail of the component's construction.